### ISSUE TITLE
[BUG] With many observables `generate_shifted_tapes()` is called "unreasonably often" resulting in massive performance loss

### ISSUE THREAD

**Issue by cvjjm** [PennyLaneAI/pennylane#2430]
### Expected behavior

When taking the parameter shift hessian of a `QNode` returning the expectation value of several compatible observables, I expect the construction of shifted tapes to only take place once per gate, then the derivatives of all observables can be computed from this single set of shifted gates.

### Actual behavior

In reality `parameter_shift.py:152:expval_param_shift` is called for each observable separately and `generate_shifted_tapes()` is therefore called number of parametrized gates times number of observables many times.

This results in a massive performance hit. In the 10 qubit example below, PennyLane spends over 60% of the time copying tapes, compared to just ~4% with the simulation of circuits, even on the slow `default.qubit` simulator.

### Additional information

To get a nice graphical overview of what is happening run the example below with `cProfile`, e.g.. like this: `python -m cProfile -o timing test.py && gprof2dot -f pstats timing -o graph.dot && dot -Tsvg graph.dot -o graph.svg`

### Source code

```shell
# test.py 
# run with:
# python -m cProfile -o timing test.py && gprof2dot -f pstats timing -o graph.dot && dot -Tsvg graph.dot -o graph.svg
import pennylane as qml
from pennylane import numpy as np

num_wires = 10
wires = range(num_wires)
dev = qml.device('default.qubit', num_wires)

np.random.seed(42)
init_params = np.random.randn(num_wires//2)

@qml.qnode(dev, diff_method='parameter-shift')
def qnode(params):
    for par_idx, wire in enumerate(wires[::2]):
        qml.Hadamard(wire)
        qml.CRY(params[par_idx], [wire, wire+1])
    return [qml.expval(qml.PauliZ(wire1) @ qml.PauliZ(wire2)) for wire1 in wires for wire2 in wires if wire1 != wire2]

print(len([0 for wire1 in wires for wire2 in wires if wire1 != wire2]))

for _ in range(4):
    print(qml.jacobian(qnode)(init_params))
```

### System information

```shell
Python 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pennylane as qml; qml.about()
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Name: PennyLane
Version: 0.21.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/cvjjm/src/covqcstack/qcware/pennylane
Requires: numpy, scipy, networkx, retworkx, autograd, toml, appdirs, semantic-version, autoray, cachetools, pennylane-lightning
Required-by: pytket-pennylane, PennyLane-Qchem, PennyLane-Lightning, covvqetools
Platform info:           Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.10
Python version:          3.8.8
Numpy version:           1.20.1
Scipy version:           1.6.1
Installed devices:
- default.gaussian (PennyLane-0.21.0.dev0)
- default.mixed (PennyLane-0.21.0.dev0)
- default.qubit (PennyLane-0.21.0.dev0)
- default.qubit.autograd (PennyLane-0.21.0.dev0)
- default.qubit.jax (PennyLane-0.21.0.dev0)
- default.qubit.tf (PennyLane-0.21.0.dev0)
- default.qubit.torch (PennyLane-0.21.0.dev0)
- pytket.pytketdevice (pytket-pennylane-0.1.0)
- lightning.qubit (PennyLane-Lightning-0.20.2)
- cov.qubit (covvqetools-0.1.1)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by cvjjm** [comment]
Here is a screenshot of the relevant part of the profiler tree:
![image](https://user-images.githubusercontent.com/60615188/162445887-8dae7df1-1f38-4892-9c09-5d4fe6b84e13.png)

**Comment by cvjjm** [comment]
Also the amount of time spend on tape hashing seems excessive :-)

**Comment by josh146** [comment]
Wow, thanks for alerting us on this @cvjjm!

> Also the amount of time spend on tape hashing seems excessive :-)

Yes, this one we are aware of and are thinking about how to fix! A solution probably involves smart usage of `lru_cache()`. We also noticed that calling of `np.round()` (as part of the caching) is _also_ a significant overhead, so something we are looking into.

It is an interesting question, since the hashing results in a significant reduction in _quantum_ resources (by reducing the amount of executions that take place), at a cost of additional classical resources. Something though that is very obvious in variational workflows!

**Comment by cvjjm** [comment]
Reducing quantum resources should always have priority imho. We want to use PL for real experiments! :-)

**Comment by mlxd** [comment]
Hi @cvjjm this is really helpful. Running your provided example takes around 40s locally. Attacking the offending function allows me to bring this down to ~27s. It may take a few days, but I'll aim to put together a general solution that helps here.

**Comment by josh146** [comment]
Thanks @cvjjm; as previously discussed it is potentially non-trivial as to _why_ this is happening, since there is no explicit for loop over observables in the parameter-shift logic 🤔

My gut feeling is that it might be related to Hamiltonian differentiation. Currently, when PL identifies that the observable coefficients might be differentiable, this function is applied: 

https://github.com/PennyLaneAI/pennylane/blob/c272db0adbe076570aab6694628423f1034bf005/pennylane/gradients/hamiltonian_grad.py#L19

which _does_ loop over observables.

So the solution could be identifying why this is applied when it isn't needed, and fixing it.

**Comment by cvjjm** [comment]
Here is some more statistics to highlight the severity of this:
- On PennyLane v0.19.0 executing the example above takes (most of this is startup time): 
```
real	0m17.312s
user	0m8.488s
sys	0m4.016s
```
- On PennyLane v0.23.0 it takes: 
```
real	2m31.671s
user	1m37.670s
sys	0m51.943s
```
The other open question was, if I understand @josh146 correctly, why `generate_shifted_tapes()` is called so often because there should be no loop over observables.

Putting into `generate_shifted_tapes()` a `traceback.print_stack()` gives some insight into what is happening:
```
  File "pl_test10.py", line 24, in <module>
    print(qml.jacobian(qnode)(init_params))
  File "/.../pennylane/pennylane/_grad.py", line 324, in _jacobian_function
    jac = tuple(_jacobian(func, arg)(*args, **kwargs) for arg in _argnum)
  File "/.../pennylane/pennylane/_grad.py", line 324, in <genexpr>
    jac = tuple(_jacobian(func, arg)(*args, **kwargs) for arg in _argnum)
  File "/.../autograd/autograd/wrap_util.py", line 20, in nary_f
    return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)
  File "/.../autograd/autograd/differential_operators.py", line 61, in jacobian
    return np.reshape(np.stack(grads), jacobian_shape)
  File "/.../autograd/autograd/numpy/numpy_wrapper.py", line 88, in stack
    arrays = [array(arr) for arr in arrays]
  File "/.../autograd/autograd/numpy/numpy_wrapper.py", line 88, in <listcomp>
    arrays = [array(arr) for arr in arrays]
  File "/.../autograd/autograd/core.py", line 14, in vjp
    def vjp(g): return backward_pass(g, end_node)
  File "/.../autograd/autograd/core.py", line 21, in backward_pass
    ingrads = node.vjp(outgrad[0])
  File "/.../autograd/autograd/core.py", line 67, in <lambda>
    return lambda g: (vjp(g),)
  File "/.../pennylane/pennylane/interfaces/autograd.py", line 188, in grad_fn
    vjp_tapes, processing_fn = qml.gradients.batch_vjp(
  File "/.../pennylane/pennylane/gradients/vjp.py", line 313, in batch_vjp
    g_tapes, fn = vjp(tape, dy, gradient_fn, gradient_kwargs)
  File "/.../pennylane/pennylane/gradients/vjp.py", line 182, in vjp
    gradient_tapes, fn = gradient_fn(tape, **gradient_kwargs)
  File "/.../pennylane/pennylane/transforms/batch_transform.py", line 330, in __call__
    return self._tape_wrapper(*targs, **tkwargs)(qnode)
  File "/.../pennylane/pennylane/transforms/batch_transform.py", line 418, in <lambda>
    return lambda tape: self.construct(tape, *targs, **tkwargs)
  File "/.../pennylane/pennylane/transforms/batch_transform.py", line 402, in construct
    tapes, processing_fn = self.transform_fn(tape, *args, **kwargs)
  File "/.../pennylane/pennylane/gradients/parameter_shift.py", line 667, in param_shift
    g_tapes, fn = expval_param_shift(tape, argnum, shifts, gradient_recipes, f0)
  File "/.../pennylane/pennylane/gradients/parameter_shift.py", line 240, in expval_param_shift
    g_tapes = generate_shifted_tapes(tape, idx, op_shifts, multipliers)
  File "/.../pennylane/pennylane/gradients/general_shift_rules.py", line 416, in generate_shifted_tapes
    traceback.print_stack(f=None, limit=None, file=None)
```
Can someone here who knows the inner guts of PL better than me make sense of this?

**Comment by mlxd** [comment]
Hi @cvjim thanks for the above. I am currently going through this process, and there are a few things going on under the hood here with the 0.23.0 release. I plan to have a solution for some of these shortly.

**Comment by josh146** [comment]
@cvjjm I'm digging into this in more detail, and I think this might be due to this issue in Autograd: https://stackoverflow.com/questions/54488875/improve-performance-of-autograd-jacobian

In particular, Autograd _itself_ has a for-loop over the output shape of the cost function it is computing the Jacobian, which is quite inefficient. As far as I can tell, this doesn't happen in Torch, TensorFlow, or JAX.

For example, the following program is _very_ fast, and doesn't have this for loop over observables:

```
import pennylane as qml
from pennylane import numpy as np
import jax
from jax import numpy as jnp

num_wires = 10
wires = range(num_wires)
dev = qml.device('default.qubit', num_wires)

np.random.seed(42)
init_params = np.random.randn(num_wires//2)

@qml.qnode(dev, diff_method='parameter-shift', interface="jax")
def qnode(params):
    for par_idx, wire in enumerate(wires[::2]):
        qml.Hadamard(wire)
        qml.CRY(params[par_idx], wires=[wire, wire+1])
    return [qml.expval(qml.PauliZ(wire1) @ qml.PauliZ(wire2)) for wire1 in wires for wire2 in wires if wire1 != wire2]

print(len([0 for wire1 in wires for wire2 in wires if wire1 != wire2]))


for _ in range(4):
    print(jax.jacobian(qnode)(init_params))
```

Then the question becomes: how come PennyLane v0.19 wasn't affected by this poor implementation of `autograd.jacobian`?

I think I can answer this: it is because, in PennyLane v0.19, we were attempting to hotfix this by caching the Jacobian _between_ autograd calls! However, when we re-wrote PennyLane for arbitrary orders of differentiation, this hotfix became a memory leak.

**Comment by josh146** [comment]
It looks like attempts were made to improve the performance of `autograd.jacobian`, but as it is now no longer maintained, this will likely never happen: https://github.com/HIPS/autograd/issues/524

**Comment by josh146** [comment]
@mlxd @cvjjm here is the former hotfix, from PR #1131:

https://github.com/PennyLaneAI/pennylane/blob/8745c6fe1188c5d3b0f406f756745fc4c3392f94/pennylane/interfaces/autograd.py#L198-L1204

It took advantage of closure in a really scary way.

**Comment by cvjjm** [comment]
Looks like a perfectly valid use of closure to me :-)

A similar "caching" effect can also be used to speed up the above example (and still get the correct result). See here for a demo: https://github.com/PennyLaneAI/pennylane/commit/408ca1decee5282e6b3ab72b0381889151fa8b2b  

![image](https://user-images.githubusercontent.com/60615188/169562875-f746a290-9c03-44d8-9499-56a24def9303.png)

This is obviously not a finished solution, one certainly needs to free the `result_dict` somehow at the right time, but it works and gives the correct results in the example above without the insane number of tape shift/copy/hash operations. 

Maybe something along those lines can be added to PL?

**Comment by josh146** [comment]
Thanks @cvjjm! I'll have a look into it. The subtlety of the previous hotfix (which I don't think is possible with the higher order diff support) is that the cache became automatically out of scope after Autograd backpropagation, which meant that Python was clearing the memory.

To replicate that, we might have to dig in and work out how to add the caching to https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/interfaces/autograd.py.

Having said that, I think rewriting `qml.jacobian` to use a mix of jvp's and vjp's -- that is more efficient than simply looping over vjp's --- is the right way to go here, and should negate the need to cache altogether

**Comment by cvjjm** [comment]
The solution you outlined with mixing of jvp's and vjp's sounds good! Is there any chance this could make it into one of the next releases?

**Comment by antalszava** [comment]
Hi @cvjjm, we have been looking into a fix to this issue in https://github.com/PennyLaneAI/pennylane/pull/2645. It would be using the caching approach. The PR needs a couple of final touches, but our plan is to have it in with the v0.25.0, in the next release. Would that work well?

**Comment by cvjjm** [comment]
Great! The implementation in #2645 looks very nice.

**Comment by antalszava** [comment]
Hi @cvjjm, we've just merged in the fix. Let us know how it is in your use case. :)

**Comment by cvjjm** [comment]
Awesome! Just tested and current master now even beats the performance of v0.180 175.19s to 250.03s in a test that took several hours to run with v0.22 - v0.24 :-)
